### ISSUE TITLE
Dynamic Topic Modeling - 

### ISSUE THREAD

**Issue by fcndata** [MaartenGr/BERTopic#1515]
Hi Maarteen,

I already fit() my model, now I would like to see the evolution of the topics over time so I follow the [steps on](https://maartengr.github.io/BERTopic/getting_started/topicsovertime/topicsovertime.html#example) but for some reason I get the error of "All arrays must be of the same length" even when they are as you can see.

Any idea of what could be wrong?

![Captura de pantalla 2023-09-08 102910](https://github.com/MaartenGr/BERTopic/assets/127878852/b22d530a-238f-4a6c-a7c1-492475341f9d)

**Comment by MaartenGr** [comment]
Could you share your full code including the full error log? It seems that the documents and timestamps you passed are not of equal length, or the same as the input for the topic model.

**Comment by fcndata** [comment]
The timestamps and the documents are the same length, as you can see. However, I trained the model with a data sampling is that the issue?

**Comment by MaartenGr** [comment]
Yes, you should pass the exact same documents as the ones you passed to either `.fit` or `.fit_transform`.

**Comment by fcndata** [comment]
So that is the reason.
@MaartenGr , I need your advice.

To give you some context, my dataset has 2MM tokens. As I want to take the full potential of my "roBERTa" embeddings, I'll use UMAP and HDBSCAN. As these two models don't allow .partial_fit to train, I thought that using a random sampling on the train would be enough to use the model later with the whole dataset, but it seems that I'm wrong as I need to use Dynamic Topic Modeling then.

Having said that: 
- What do you suggest?
- Is a V100 GPU from collab enough to train the full model or should I go for the A100 GPU?

**Comment by MaartenGr** [comment]
Sure, let's at the beginning though. You are referring to "roBERTa" embeddings. Which embeddings did you mean by that? Also, how many documents do you have? Lastly, could you share your full code?

**Comment by fcndata** [comment]
With "roBERTa"  I mean the version "all-distilroberta-v1" on SentenceTransformer and the documents are 1.585.391 and you can see them on the notebook as I printed them.

Here you can find [my notebook](https://we.tl/t-AOYch8iHiq)

**Comment by MaartenGr** [comment]
> With "roBERTa" I mean the version "all-distilroberta-v1" on SentenceTransformer and the documents are 1.585.391 and you can see them on the notebook as I printed them.

Alright, so generally, I would advise not going with that model but using something like "all-mpnet-base-v2" instead. You can find an overview [here](https://www.sbert.net/docs/pretrained_models.html) or [here](https://huggingface.co/spaces/mteb/leaderboard). With that many documents, I would highly advise following the notebook on how to use BERTopic with large data [here](https://github.com/MaartenGr/BERTopic#getting-started). 

> Here you can find [my notebook](https://we.tl/t-AOYch8iHiq)

Thanks but I do not download external files to my local computer this way. If its hosted on Github or Google Drive, I can take a look. Otherwise, you can share your code here.

**Comment by MaartenGr** [comment]
With respect to the GPU. You should actually get quite far with the T4 already, so using the V100 should definitely suffice! But make sure to follow the tutorial I mentioned above for more detailed instructions.

**Comment by fcndata** [comment]
Thanks for your comments; I'll look at the tutorials you were talking about.

Here is the [notebook](https://colab.research.google.com/drive/1h_K9O1pakcUlMCoLA0di4nqB5L6t63TO?usp=drive_link) on Drive so you can take a look would be great!

Thanks Maarten!

**Comment by MaartenGr** [comment]
Based on your notebook, I would indeed advise starting with the notebook I mentioned before on using BERTopic with large data. Start with those settings and parameters as I think they might be better suited for your application. For instance, you have set `n_components` to 35 which is quite a lot for clustering tasks and you might get better results with lower values. So stick with that example notebook first before you start hyperparameter tuning. I would advise keeping HDBSCAN model as is since it seems you found a good `min_cluster_size`. Also, definitely do not set `calculate_probabilities=True` when using the default HDBSCAN. It uses significant memory and can be quite slow.

**Comment by fcndata** [comment]
Maarten,

First of all, amazing job [here](https://github.com/MaartenGr/BERTopic#getting-started).

In 'all-mpnet-base-v2', the "Max Sequence Length" is 384, in that case, should I check the size tokens?

I thought that `n_components` it was the dimensionality in which you condense the 700+ dimensions from the embeddings, that is why I set 35 to don't overcompress the data.

Thanks!

**Comment by MaartenGr** [comment]
There is a fine balance between compressing the data and preventing the curse of dimensionality. I would advise starting first with default values and then building up from there. Going from 5 to 35 is quite the leap and may also prove to be computationally difficult. 

With respect tot the token size, that depends on how much your documents exceed the token size. I would start just training and then see if anything is missing or not working correctly. You would be surprised with how well it might work out of the box and following the best practices and notebook for large data.

**Comment by fcndata** [comment]
Maarten, what do you think about this: 

![image](https://github.com/MaartenGr/BERTopic/assets/127878852/7978bd51-4522-4490-85c3-14b2bacb6f2d)

These are my topics, and to be honest, it makes sense. 

My goal is to understand the evolution of the perception of AI, to do so, I trained the model, to visualise how different AI events are impacted. However, as everything is related to "AI", most of the documents are in Topic 0 (great to see how the different languages also are on one different topic). Having said that, what do you suggest for me to get deeper on topic 0.

**Comment by MaartenGr** [comment]
It depends, can you share your exact code for getting to these specific topics? How did you train it and with what parameters? Did you already use the `all-mpnet-base-v2` or something else? Did you reduce to 5 dimensions or did you keep the 35 dimensions you mentioned earlier?

**Comment by fcndata** [comment]
No, different parameters, mb. 
Now I took as a baseline what you suggested me:

`embedding_model = SentenceTransformer('sentence-transformers/all-mpnet-base-v2')
umap_model = UMAP(n_components=5, n_neighbors=50, random_state=42, metric="cosine", verbose=True)
hdbscan_model =HDBSCAN(min_samples=20, gen_min_span_tree=True, prediction_data=False, min_cluster_size=20, verbose=True)
vectorizer_model = CountVectorizer(vocabulary=vocab,ngram_range=(1, 2),min_df=3,stop_words="english")
ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True,bm25_weighting=True)
keybert_model = KeyBERTInspired()
mmr_model = MaximalMarginalRelevance(diversity=0.3)`

These are the results, they look much better:
![image](https://github.com/MaartenGr/BERTopic/assets/127878852/e44e7191-2694-4f2b-a189-24dc044cae5c)

Now, I can see how different topics are expressed over time: 
![image](https://github.com/MaartenGr/BERTopic/assets/127878852/c3ffcc06-eb19-4f06-9bc8-6d94341da477)

I was thinking of using the [Topic Modeling with Llama2](https://colab.research.google.com/drive/1QCERSMUjqGetGGujdrvv_6_EeoIcd_9M?usp=sharing#scrollTo=w1ufudzoUZzH) to the definition of the topics, but is it possible to use it for the evolution tuning for _Dynamic Topic Modeling_

`topic_model.topics_over_time(docs=docum, 
                                                timestamps=timestamps, 
                                                global_tuning=True, 
                                                evolution_tuning=True, 
                                                nr_bins=13)`

Thanks Maarten!

**Comment by MaartenGr** [comment]
> I was thinking of using the [Topic Modeling with Llama2](https://colab.research.google.com/drive/1QCERSMUjqGetGGujdrvv_6_EeoIcd_9M?usp=sharing#scrollTo=w1ufudzoUZzH) to the definition of the topics, but is it possible to use it for the evolution tuning for Dynamic Topic Modeling

That is currently not possible since topic aspects and different representations can computationally take quite a while to run if you perform dynamic topic modeling. Dynamic topic modeling is essentially running the representations *n* times, one for each timestamp. So scaling that can be difficult. 

> Another question: is it possible to use the embedding space of each topic and Llama2, to define if some topics should be merged?

Yes, that is what is happening in the visualization that you show. Because their embeddings are similar, they might be able to merge.
